juju/applications.go

```diff
@@ -0,0 +1,560 @@
+	}
+	resolvedCharm := resolved[0]
+
+	if resolvedCharm.Error != nil {
+		log.Info("Error when resolving charm", "charm", resolvedCharm)
+		return nil, resolvedCharm.Error
+	}
+
+	// Add the charm to the model
+	origin = resolvedCharm.Origin.WithBase(&base)
+
+	var deployRevision int
+	if input.CharmRevision > -1 {
+		deployRevision = input.CharmRevision
+	} else {
+		if origin.Revision != nil {
+			deployRevision = *origin.Revision
+		} else {
+			return nil, errors.New("no origin revision")
+		}
+	}
+
+	charmURL = resolvedCharm.URL.WithRevision(deployRevision).WithArchitecture(origin.Architecture).WithSeries(series)
+	resultOrigin, err := charmsAPIClient.AddCharm(charmURL, origin, false)
+	if err != nil {
+		return nil, err
+	}
+
+	charmID := apiapplication.CharmID{
+		URL:    charmURL,
+		Origin: resultOrigin,
+	}
+
+	resources, err := c.processResources(charmsAPIClient, resourcesAPIClient, charmID, input.ApplicationName)
+	if err != nil {
+		return nil, err
+	}
+
+	// The deploy API endpoint expects string values for the
+	// constraints.
+	var appConfig map[string]string
+	if input.Config == nil {
+		appConfig = make(map[string]string)
+	} else {
+		appConfig = make(map[string]string, len(input.Config))
+		for k, v := range input.Config {
+			appConfig[k] = ConfigEntryToString(v)
+		}
+	}
+
+	appConfig["trust"] = fmt.Sprintf("%v", input.Trust)
+
+	err = applicationAPIClient.Deploy(apiapplication.DeployArgs{
+		CharmID:         charmID,
+		ApplicationName: appName,
+		NumUnits:        input.Units,
+		CharmOrigin:     resultOrigin,
+		Config:          appConfig,
+		Cons:            input.Constraints,
+		Resources:       resources,
+	})
+
+	if err != nil {
+		// unfortunate error during deployment
+		return &CreateApplicationResponse{
+			AppName:  appName,
+			Revision: *origin.Revision,
+			Base:     origin.Base.String(),
+		}, err
+	}
+
+	return &CreateApplicationResponse{
+		AppName:  appName,
+		Revision: *origin.Revision,
+		Base:     origin.Base.String(),
+	}, err
+}
+
+// processResources is a helper function to process the charm
+// metadata and request the download of any additional resource.
+func (c applicationsClient) processResources(charmsAPIClient *apicharms.Client, resourcesAPIClient *apiresources.Client, charmID apiapplication.CharmID, appName string) (map[string]string, error) {
+	charmInfo, err := charmsAPIClient.CharmInfo(charmID.URL.String())
+	if err != nil {
+		return nil, err
+	}
+
+	// check if we have resources to request
+	if len(charmInfo.Meta.Resources) == 0 {
+		return nil, nil
+	}
+
+	pendingResources := []charmresources.Resource{}
+	for _, v := range charmInfo.Meta.Resources {
+		aux := charmresources.Resource{
+			Meta: charmresources.Meta{
+				Name:        v.Name,
+				Type:        v.Type,
+				Path:        v.Path,
+				Description: v.Description,
+			},
+			Origin: charmresources.OriginStore,
+			// TODO: prepare for resources with different versions
+			Revision: -1,
```

Comment:
What happens when the resource revisions are set to `-1` here? Is the controller going to fetch the right resource for the 1.26/stable channel?

Reply:
-1 is essentially saying give me whatever charmhub has as the resource associated with the channel. That happens [here](https://github.com/juju/juju/blob/117f63ec1e97a31ddedc92aa1af1f32eb378d718/apiserver/facades/client/resources/repository.go#L211)